src/checkers/firefoxchecker.py

```diff
@@ -0,0 +1,181 @@
+        log.debug('Retrieving latest available firefox version')
+        try:
+            self._get_latest_available_version()
+        except Exception as e:
+            log.warning('Error retrieving latest firefox version: %s', e)
+        finally:
+            if not self._latest_version:
+                log.warning('Could not determine latest firefox version')
+                return None
+
+        log.debug('Latest available firefox version: %s', self._latest_version)
+
+        log.debug('Retrieving latest firefox version info')
+        latest_info = None
+        try:
+            latest_info = self._get_latest_version_info(self._latest_version, checker_data)
+        except Exception as e:
+            log.warning('Error retrieving latest firefox version info: %s', e)
+        finally:
+            if not latest_info:
+                log.warning('Could not retrieve latest firefox version information')
+                return None
+
+        log.debug('Checking external data')
+        browser_data = None
+        for data in external_data:
+            if re.match(checker_data['browser']['source-filename'], data.filename):
+                browser_data = data
+                break
+
+        if not browser_data:
+            log.warning('Invalid x-checker-data for module %s, unable to find browser source with filename %s',
+                        module_data.name,
+                        checker_data['browser']['source-filename'])
+            return None
+
+        added = []
+        processed_data = []
+        for data in external_data:
+            if data.filename in latest_info:
+                new_version = latest_info[data.filename]
+                data.state = ExternalData.State.VALID
+                if not data.current_version.matches(new_version):
+                    data.new_version = new_version
+            else:
+                if re.match(checker_data['browser']['filename-match'], data.filename) or \
+                   re.match(checker_data['translations']['filename-match'], data.filename):
+                    data.state = ExternalData.State.REMOVED
+            processed_data.append(data.filename)
+
+        added = []
+        new_sources = []
+        for source_filename, info in latest_info.items():
+            if not source_filename in processed_data:
+                source = {
+                    'filename': source_filename,
+                    'type': 'extra-data',
+                    'url': info.url,
+                    'sha256': info.checksum,
+                    'size': info.size,
+                }
+                # add new data to the same manifest as the main browser tarball
+                data = ExternalDataSource.from_source(browser_data.source_path, source, browser_data.source_parent)
+                data.state = ExternalData.State.ADDED
+                added.append(data)
+
+        return added
+
+    def _get_latest_available_version(self):
+        if self._latest_version:
```

Comment:
I think you can drop all the manual caching and decorate this function with [functools.lru_cache()](https://docs.python.org/3/library/functools.html#functools.lru_cache):

```python3
    @functools.lru_cache()
    def _get_latest_available_version(self):
```

Reply:
Ok, but isn't this more expensive than the current cache? Not that we care much about resources here but still...

Reply:
It'll be a dictionary whose key is `self` and whose value is the return value – a dozen bytes or so. I think this is worth the code size reduction.

Reply:
Done.